src/typesystem.hpp

```diff
@@ -67,4 +62,26 @@ struct TypeSystemCore {
 	MonoId inst_impl(MonoId mono, std::unordered_map<MonoId, MonoId> const& mapping);
 	MonoId inst_with(PolyId poly, std::vector<MonoId> const& vals);
 	MonoId inst_fresh(PolyId poly);
+
+	TypeFunctionData& type_function_data_of(MonoId);
+	TypeFunctionId new_tf_var();
+	void unify_tf(TypeFunctionId, TypeFunctionId);
+private:
+	int compute_new_argument_count(
+	    TypeFunctionData const& a_data, TypeFunctionData const& b_data) const;
+
+	void point_tf_at_another(TypeFunctionId a, TypeFunctionId b);
+	void unify_tf_data(TypeFunctionData& a_data, TypeFunctionData& b_data);
+	TypeFunctionData& get_tf_data(TypeFunctionId tf);
+	TypeFunctionId find_tf(TypeFunctionId tf);
+	TypeFunctionId create_tf(
+	    TypeFunctionTag tag,
+	    int arity,
+	    std::vector<InternedString> fields,
+	    std::unordered_map<InternedString, MonoId> structure,
+	    bool is_dummy);
+
+	std::vector<TypeFunctionData> m_type_functions;
+	std::vector<PolyData> poly_data;
+	UnionFind m_tf_uf;
```

Comment:
Nomenclature is all over the place. We use "tf" and "type_function" interchangeably, we should pick one.

Reply:
Yeah, which one do you like? Reading code, `tf` is immediately recognizable, but `type_function` is just nice to see innit? I used `tf` because it was less typing

Reply:
`type_function` is best in my opinion. The names of the functions need to be inversely proportional to the scope in which they live, as Uncle Bob stated.

Reply:
I don't think that's relevant